FreeRTOS/Demo/ARC_NSIM_QEMU_GCC_MW/arc/board.c

```diff
@@ -0,0 +1,220 @@
+/* ------------------------------------------
+ * Copyright (c) 2022, Synopsys, Inc. All rights reserved.
+
+ * Redistribution and use in source and binary forms, with or without modification,
+ * are permitted provided that the following conditions are met:
+
+ * 1) Redistributions of source code must retain the above copyright notice, this
+ * list of conditions and the following disclaimer.
+
+ * 2) Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation and/or
+ * other materials provided with the distribution.
+
+ * 3) Neither the name of the Synopsys, Inc., nor the names of its contributors may
+ * be used to endorse or promote products derived from this software without
+ * specific prior written permission.
+
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+   --------------------------------------------- */
+#include "arc.h"
+#include "board.h"
+#include "arc_timer.h"
+#include "arc_exception.h"
+#include <stdint.h>
+#include <stddef.h>
+#include <stdio.h>
+#include <ctype.h>
+#include <errno.h>
+
+typedef struct main_args {
+	int argc;
+	char *argv[];
+} MAIN_ARGS;
+
+/** Change this to pass your own arguments to main functions */
+MAIN_ARGS s_main_args = { 1, { "main" } };
+
+/** board timer interrupt reset count */
+static uint32_t cyc_hz_count = (BOARD_CPU_CLOCK / BOARD_SYS_TIMER_HZ);
+
+/** board timer counter in timer interrupt */
+static volatile uint64_t gl_sys_hz_cnt = 0;
+/** board 1ms counter */
+static volatile uint32_t gl_ms_cnt = 0;
+
+#define HZ_COUNT_CONV(precision, base)  ((precision) / (base))
+
+extern void __mw_run_tls_dtor();
+
+void task_end_hook( void * pxTCB )
+{
+	TaskHandle_t task2free = ( TaskHandle_t ) pxTCB;
+
+	if( task2free != NULL )
+	{
+		void * tls = pvTaskGetThreadLocalStoragePointer( task2free, 0 );
+
+		if( tls )
+		{
+			__mw_run_tls_dtor();
+			vPortFree( tls );
+			vTaskSetThreadLocalStoragePointer( task2free, 0, NULL );
+		}
+	}
+}
+
+void gnu_printf_setup(void)
+{
+#if defined(__GNU__) && !defined(_HAVE_LIBGLOSS_) && !defined(_HOSTLINK_)
+	exc_handler_install(EXC_NO_SWI, (EXC_HANDLER_T)syscall_swi);
```

Comment:
Hmm, any explanation about why do we need this?

We don't need SWI exception for printf, I suppose.

Reply:
You're right, we actually don't need SWI, but we still need `gnu_printf_setup()`, which is defined in previous 
 [FreeRTOS ARC support](https://github.com/FreeRTOS/FreeRTOS-Kernel/blob/d5a10e45958148d437ae5096835a118be58e6df9/portable/ThirdParty/GCC/ARC_EM_HS/arc_freertos_exceptions.c#:~:text=arc_freertos_exceptions.h%22-,%23ifdef%20__GNU__,%23endif,-/**) . So I left this function blank and have a comment on top of it.